### PR TITLE
[RISCV] Support constraint "s"

### PR DESCRIPTION
**Pull request by MaskRay**
GCC has supported a generic constraint "s" for a long time (since at
least 1992), which references a symbol or label with an optional
constant offset. "i" is a superset that also supports a constant
integer.

GCC's RISC-V port also supports a machine-specific constraint "S",
which cannot be used with a preemptible symbol. (We don't bother to
check preemptibility.) In PIC code, an external symbol is preemptible by
default, making "S" less useful if you want to create an artificial
reference for linker garbage collection, or define sections to hold
symbol addresses:

```
void fun();
// error: impossible constraint in ‘asm’ for riscv64-linux-gnu-gcc -fpie/-fpic
void foo() { asm(".reloc ., BFD_RELOC_NONE, %0" :: "S"(fun)); }
// good even if -fpie/-fpic
void foo() { asm(".reloc ., BFD_RELOC_NONE, %0" :: "s"(fun)); }
```

This patch adds support for "s". Modify https://reviews.llvm.org/D105254
("S") to handle multi-depth GEPs (https://reviews.llvm.org/D61560).

### CODE REVIEW

**Comment by MaskRay** [comment]
The discussion with GCC seems to suggest we should encourage "s" for metadata section usage.

I have some notes about these constraints https://maskray.me/blog/2024-01-30-raw-symbol-names-in-inline-assembly 

> While the default implementation (gcc/defaults.h) is permissive (used by MIPS, PowerPC, and RISC-V), many ports impose stricter restrictions, often disallowing preemptible symbols under PIC.
> ...
> Nevertheless, I think this symbol preemptibility limitation for "s" is unfortunate. Ideally, we could retain the current "i" for immediate integer operand (after linking), and design "s" for a raw symbol name with a constant offset, ignoring symbol preemptibility. This architecture-agnostic "s" would simplify metadata section utilization and boost code portability.

File: llvm/test/CodeGen/RISCV/inline-asm-s-constraint.ll [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+@var = external dso_local global i32, align 4
+@a = external global [2 x [2 x i32]], align 4
+
+define dso_local void @test() {
+; CHECK-LABEL: test:
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    #APP
+; CHECK-NEXT:    # var a+12 test
+; CHECK-NEXT:    #NO_APP
+; CHECK-NEXT:    ret{{[l|q]}}
```
Comment (kito-cheng) [inline llvm/test/CodeGen/RISCV/inline-asm-s-constraint.ll:14]: I don't see any line check with `CHECK`? maybe you forgot to drop after regen?

**Review by kito-cheng — APPROVED** [review]
Just one minor comment, otherwise LGTM :)

File: llvm/test/CodeGen/RISCV/inline-asm-s-constraint.ll (hunk 1 above)
Comment (MaskRay) [reply llvm/test/CodeGen/RISCV/inline-asm-s-constraint.ll]: Thanks for catching this. Removed

**Pull request merged**